backend/internal/handlers/utils.go

```diff
@@ -0,0 +1,81 @@
+package handlers
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/golang/gddo/httputil/header"
+)
+
+type malformedRequest struct {
+	status int
+	msg    string
+}
+
+func (mr *malformedRequest) Error() string {
+	return mr.msg
+}
+
+// From https://www.alexedwards.net/blog/how-to-properly-parse-a-json-request-body
+func decodeJSONBody(w http.ResponseWriter, r *http.Request, dst interface{}) error {
+	if r.Header.Get("Content-Type") != "" {
+		value, _ := header.ParseValueAndParams(r.Header, "Content-Type")
+		if value != "application/json" {
+			msg := "Content-Type header is not application/json"
+			return &malformedRequest{status: http.StatusUnsupportedMediaType, msg: msg}
+		}
+	}
+
+	r.Body = http.MaxBytesReader(w, r.Body, 1048576)
+
+	dec := json.NewDecoder(r.Body)
+	dec.DisallowUnknownFields()
+
+	err := dec.Decode(&dst)
+	if err != nil {
+		var syntaxError *json.SyntaxError
```

Comment:
This error message handling seems redundant, just wonder if the Decode error text is too encryptic that it requires additional handling.

Hm...would this be enough? 

```golang
if err != nil {
   return &malformedRequest{status: http.StatusBadRequest, msg: err}
}
```

Reply:
Valid question, the author's justification in the [linked blog](https://www.alexedwards.net/blog/how-to-properly-parse-a-json-request-body) is

> The error messages returned by Decode() aren't ideal for sending to a client. Some are arguably too detailed and expose information about the underlying program (like "json: cannot unmarshal number into Go struct field Person.Name of type string"). Others aren't descriptive enough (like "unexpected EOF") and some are just plain confusing (like "invalid character 'A' looking for beginning of object key string"). There also isn't consistency in the formatting or language used.

which makes sense to me. I wouldn't go through the trouble of writing this all out myself for this project, but since this guy has already done most of the heavy lifting and published it for free, I figure I may as well take the help.

Reply:
I would say that saying a client (in this case web ui) does not need to know these details, some generic message "Unable to process request" would suffice and we can log detailed Decode error message as is on the server side. 

If API itself was a product (as `stripe` for example) then I would of course give more attention to error messages returning to the client, but in this particular case it is not worth it. 

For example, would we want to show a user this message in UI?
```
Request body contains an invalid value for the "xxx" field (at position 454)
```

Yeah, lets remove this for now (i do agree that Decode messages might be hard to read but I would not try to optimize it). 

One thing that I do not like is that the author logic will break if internals of the Decode method change, like this string comparsion `strings.HasPrefix(err.Error(), "json: unknown field ")`